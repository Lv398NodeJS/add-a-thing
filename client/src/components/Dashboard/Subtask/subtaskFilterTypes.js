const SHOW_ALL = 'SHOW_ALL';
const SHOW_ACTIVE = 'SHOW_ACTIVE';
const SHOW_COMPLETED = 'SHOW_COMPLETED';

export {
  SHOW_ALL,
  SHOW_ACTIVE,
  SHOW_COMPLETED,
};
