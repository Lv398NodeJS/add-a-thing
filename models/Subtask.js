const mongoose = require('mongoose');

const { Schema } = mongoose;
const SubtaskSchema = new Schema({
  name: {
    type: String,
  },
  completed: Boolean,
  taskId: {
    type: Schema.Types.ObjectId,
    ref: 'Task',
    required: true,
  },
  date: {
    type: Date,
    default: Date.now,
  },
});

module.exports = Subtask = mongoose.model('Subtask', SubtaskSchema);
