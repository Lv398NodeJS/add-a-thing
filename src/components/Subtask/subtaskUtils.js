import { SHOW_ALL, SHOW_ACTIVE, SHOW_COMPLETED } from './subtaskFilterTypes';

export const getSubtaskListAsArray = (snapValue = {}) => {
  if (typeof snapValue !== 'object' && snapValue.constructor !== Object) return [];
  const subtaskList = [];
  Object.keys(snapValue).forEach((subtask) => {
    const { text, completed } = snapValue[subtask];
    subtaskList.push({
      text,
      completed,
      id: subtask,
    });
  });
  return subtaskList;
};

export const getVisibleSubtasks = (subtaskList, currentFilter) => {
  switch (currentFilter) {
    case SHOW_ALL:
      return subtaskList;
    case SHOW_COMPLETED:
      return subtaskList.filter(subtask => subtask.completed);
    case SHOW_ACTIVE:
      return subtaskList.filter(subtask => !subtask.completed);
    default:
      throw new Error(`Unknown filter: ${currentFilter}`);
  }
}