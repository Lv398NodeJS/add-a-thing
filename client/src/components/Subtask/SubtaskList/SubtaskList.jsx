import React, { Component } from 'react';
import { connect } from 'react-redux';
import SubtaskItemContainer from '../SubtaskItem/SubtaskItemContainer';
import { getVisibleSubtasks } from '../subtaskUtils';

class SubtaskList extends Component {
  render() {
    const { taskStatus, subtaskList = {}, currentFilter } = this.props;
    const subtaskItems = getVisibleSubtasks(subtaskList, currentFilter).map(subtask => (
      <SubtaskItemContainer
        key={subtask._id}
        _id={subtask._id}
        name={subtask.name}
        completed={subtask.completed}
        taskStatus={taskStatus}
      />
    ));

    return (subtaskItems);
  }
}

const mapStateToProps = ({ subtaskReducer: { taskStatus, subtaskList, currentFilter } }) => ({
  taskStatus,
  subtaskList,
  currentFilter,
});

export { SubtaskList as SubtaskListComponent };
export default connect(
  mapStateToProps,
  null,
)(SubtaskList);
