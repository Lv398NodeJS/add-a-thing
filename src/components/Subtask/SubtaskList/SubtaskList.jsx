import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import SubtaskItem from '../SubtaskItem/SubtaskItem';
import * as subtaskActions from '../../../actions/subtaskActions';
import { getVisibleSubtasks } from '../subtaskUtils';

class SubtaskList extends Component {
  constructor(props) {
    super(props);
    const {
      taskRef,
      subtaskListActions: { fetchInfoForSubtaskList },
    } = this.props;
    fetchInfoForSubtaskList(taskRef);
  }

  render() {
    const {
      taskRef, taskStatus, subtaskList = {}, currentFilter,
    } = this.props;
    const subtaskItems = getVisibleSubtasks(subtaskList, currentFilter).map(subtask => (
      <SubtaskItem
        key={subtask.id}
        id={subtask.id}
        text={subtask.text}
        completed={subtask.completed}
        taskStatus={taskStatus}
        taskRef={taskRef}
      />
    ));

    return (subtaskItems);
  }
}

const mapStateToProps = ({ subtaskReducer: { taskStatus, subtaskList, currentFilter } }) => ({
  taskStatus,
  subtaskList,
  currentFilter,
});

const mapDispatchToProps = dispatch => ({
  subtaskListActions: bindActionCreators(subtaskActions, dispatch),
});

export { SubtaskList as SubtaskListComponent };
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SubtaskList);
