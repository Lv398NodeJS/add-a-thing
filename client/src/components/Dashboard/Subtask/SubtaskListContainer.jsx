import React, { Component } from 'react';
import { Row, Col, Container } from 'react-bootstrap';
import { connect } from 'react-redux';
import SubtaskProgressBar from './SubtaskProgressBar/SubtaskProgressBar';
import SubtaskList from './SubtaskList/SubtaskList';
import SubtaskFilterContainer from './SubtaskFilter/SubtaskFilterContainer';
import SubtaskAdd from './SubtaskAdd/SubtaskAdd';

export class SubtaskListContainer extends Component {
  render() {
    const {
      taskId,
      taskStatus,
      subtaskList = {},
      currentFilter,
    } = this.props;
    return (
      <Container fluid className="my-0 mx-0">
        <Row>
          <Col className="col-4">
            <h5>Subtasks:</h5>
          </Col>
          <Col className="col-8 text-right">
            <SubtaskFilterContainer />
          </Col>
        </Row>
        <SubtaskProgressBar subtaskList={subtaskList} />
        <SubtaskList
          taskId={taskId}
          taskStatus={taskStatus}
          subtaskList={subtaskList}
          currentFilter={currentFilter}
        />
        <SubtaskAdd taskId={taskId} taskStatus={taskStatus} />
      </Container>
    );
  }
}

const mapStateToProps = (
  {
    subtaskReducer: { subtaskList, currentFilter },
    taskDetailsReducer: { taskDetails: { _id: taskId, status: taskStatus } },
  },
) => ({
  subtaskList, currentFilter, taskId, taskStatus,
});

export { SubtaskListContainer as SubtaskListContainerComponent };
export default connect(
  mapStateToProps,
  null,
)(SubtaskListContainer);
