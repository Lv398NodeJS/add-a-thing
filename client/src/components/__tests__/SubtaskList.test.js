import React from 'react';
import { shallow } from 'enzyme';
import { SubtaskListComponent } from '../Subtask/SubtaskList/SubtaskList';
import { SHOW_ALL } from '../Subtask/subtaskFilterTypes';

const subtaskList = [
  { completed: false, _id: '1', name: 'Test subtask 1' },
  { completed: true, _id: '2', name: 'Test subtask 2' },
  { completed: false, _id: '3', name: 'Test subtask 3' },
];
const subtaskListComponent = shallow(
  <SubtaskListComponent
    subtaskList={subtaskList}
    taskStatus="To Do"
    currentFilter={SHOW_ALL}
  />,
);

describe('SubtaskList component', () => {
  it('should render correctly and match the snapshot', () => {
    expect(subtaskListComponent).toMatchSnapshot();
  });

  it('should render proper count of SubtaskItemContainer components', () => {
    expect(subtaskListComponent.find('SubtaskItemContainer').length).toBe(3);
  });
});
