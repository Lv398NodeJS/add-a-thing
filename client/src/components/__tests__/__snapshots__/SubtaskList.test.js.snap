// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtaskList component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubtaskList
    currentFilter="SHOW_ALL"
    subtaskActions={
      Object {
        "fetchInfoForSubtaskList": [MockFunction] {
          "calls": Array [
            Array [
              undefined,
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    subtaskList={
      Array [
        Object {
          "completed": false,
          "id": "1",
          "text": "Test subtask 1",
        },
        Object {
          "completed": false,
          "id": "2",
          "text": "Test subtask 2",
        },
        Object {
          "completed": false,
          "id": "3",
          "text": "Test subtask 3",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "function",
    "props": Object {
      "completed": false,
      "id": "1",
      "taskRef": undefined,
      "taskStatus": undefined,
      "text": "Test subtask 1",
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(SubtaskItem)",
      "type": [Function],
      Symbol(Symbol.hasInstance): [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "function",
      "props": Object {
        "completed": false,
        "id": "1",
        "taskRef": undefined,
        "taskStatus": undefined,
        "text": "Test subtask 1",
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(SubtaskItem)",
        "type": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
    Object {
      "instance": null,
      "key": "2",
      "nodeType": "function",
      "props": Object {
        "completed": false,
        "id": "2",
        "taskRef": undefined,
        "taskStatus": undefined,
        "text": "Test subtask 2",
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(SubtaskItem)",
        "type": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
    Object {
      "instance": null,
      "key": "3",
      "nodeType": "function",
      "props": Object {
        "completed": false,
        "id": "3",
        "taskRef": undefined,
        "taskStatus": undefined,
        "text": "Test subtask 3",
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(SubtaskItem)",
        "type": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
